### ISSUE TITLE
(Philosophical question) Make dependency more explicit?

### ISSUE THREAD

**Issue by fzyzcjy** [mobxjs/mobx.dart#937]

**Comment by MaoHolden** [comment]
This reminds me of Cerebral.js, they have a Single State with arbitrary depth and you specify which parts of the state you want to listen to:

```js
import React from 'react'
import { state, sequences } from 'cerebral'
import { connect } from '@cerebral/react'

export default connect(
  {
    foo: state`foo`,
    onClick: sequences`onClick`
  },
  class MyComponent extends React.Component {
    render() {
      return <div onClick={() => this.props.onClick()}>{this.props.foo}</div>
    }
  }
)
```

Its very nice, very intuitive and specific IMO. But maybe make it optional? In small and medium projects just using the builder without worrying what store do you need is very nice.

I have a store of stores with something resembling a Single State so I just do:

```dart
        Observer(
          builder: (context) {
            if (store.wallet.withdraw.isFetchingWithdrawalInfo) {
              return const CustomProgressIndicator();
            }
            // if (store.otherStore.someObservable) ...
            
            return Text('${withdraw.userBalanceShib}');
          },
        )
``` 

I'm keeping it as simple as possible, so I would argue that if this change is introduced would be nice to make it optional. It adds too much code to my simple app. :)

**Comment by MaoHolden** [comment]
@fzyzcjy what about creating a linter rule to avoid using observables without Observer( inside a widget 🤔

**Comment by fzyzcjy** [comment]
@MaoHolden 

> what about creating a linter rule to avoid using observables without Observer( inside a widget

That will be quite hard indeed :/ It is possible to call a normal function, which calls a normal function, etc, and finally deep nested it fetches an observable.

**Comment by MaoHolden** [comment]
@fzyzcjy could this be related to that?
```
always: If observables are read outside actions/reactions, throw an Exception never: Allow unrestricted reading of observables everywhere. This is the default.
```
https://pub.dev/documentation/mobx/latest/mobx/ReactiveReadPolicy.html

**Comment by MaoHolden** [comment]
Looks like the reactive context has policies for when you can read/write observables, and looks like its possible to tell mobx to throw an error if an observable was read outside a reactive context, like an action or a reaction.

https://mobx.netlify.app/api/context/

**Comment by MaoHolden** [comment]
@fzyzcjy just realized you are contributor lol, you already know all of these things. 🙈

**Comment by fzyzcjy** [comment]
Yes, I know that. That one is great, but it is caught at runtime (instead of compile time), and there is no syntax hint to show explicit dependency (whether explicit dependency is a burden or a help is debatable!).

Again, this thread is a brainstorm, instead of any actual proposal to really implement something :) I just have something in my mind and thus output it here in case it is interesting for anyone.

**Comment by MaoHolden** [comment]
Got it, so Reactive Policies are enforced at runtime, that's not on the docs, good to know.

Then it definitely ads more safety, probably would also make mocking tests for individual widgets very easy since you know the store dependencies beforehand right?

**Comment by pavanpodila** [comment]
We have to spend time on improving the documentation. I wrote the first version few years back and need help to revisit and make it more comprehensive.

**Comment by MaoHolden** [comment]
@pavanpodila I'm reading the codebase and trying to make sense of it, after I get an idea of how this works I'll try to do some PRs for the docs. Is there anywhere you discuss documentation stuff?

**Comment by pavanpodila** [comment]
No location yet but we can start a separate thread. Look forward to your PRs

**Comment by subzero911** [comment]
> So a very very naive change to mobx may be: Add ref as an argument to Observer(builder: (context, ref) => ...). For https://github.com/observable, maybe enhance the codegen such that it gives some reader object that can be used like ref.watch(store.myfield). For @computed, maybe something similar (since again we use codegen).

@fzyzcjy 
It's great. But Observers' advantage is to track several observables at once.
With `ref` it wouldn't be possible anymore.

**Comment by subzero911** [comment]
@fzyzcjy you can save the current behaviour, just rename current Observer into AutoObserver. And add a breaking change note, that existing projects should now find-and-replace all Observers to AutoObservers in their codebases.
And also add Observer, Observer2, Observer3... Observer8, for explicit depencencies. It would be similar to https://pub.dev/documentation/provider/latest/provider/Selector-class.html
